crates/sql/src/ast/subscription.rs

```diff
@@ -0,0 +1,258 @@
+//!     ;
+//!
+//! op
+//!     = '='
+//!     | '<'
+//!     | '>'
+//!     | '<' '='
+//!     | '>' '='
+//!     ;
+//!
+//! literal
+//!     = INTEGER
+//!     | STRING
+//!     | HEX
+//!     | TRUE
+//!     | FALSE
+//!     ;
+//! ```
+//!
+use crate::ast::*;
+use crate::errors::{SqlError, SqlUnsupported};
+
+/// The maximum number of `JOIN` clauses allowed in a `SQL` subscription query.
+///
+/// This is a limitation to avoid complex queries that could be hard to optimize.
+pub const MAX_JOINS: usize = 2;
+
+/// The comparison operators supported in `SQL` subscription queries.
+///
+/// ```ebnf
+/// op
+///     = '='
+///     | '<'
+///     | '>'
+///     | '<' '='
+///     | '>' '='
+///     ;
+/// ```
+#[derive(Debug, Copy, Clone, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum SqlOpCmp {
+    Eq,
+    Lt,
+    LtEq,
+    Gt,
+    GtEq,
+}
+
+impl TryInto<SqlOpCmp> for OpCmp {
+    type Error = SqlError;
+
+    fn try_into(self) -> Result<SqlOpCmp, Self::Error> {
+        match self {
+            OpCmp::Eq => Ok(SqlOpCmp::Eq),
+            OpCmp::Lt => Ok(SqlOpCmp::Lt),
+            OpCmp::LtEq => Ok(SqlOpCmp::LtEq),
+            OpCmp::Gt => Ok(SqlOpCmp::Gt),
+            OpCmp::GtEq => Ok(SqlOpCmp::GtEq),
+            _ => Err(SqlUnsupported::SubscriptionNotEq.into()),
+        }
+    }
+}
+
+/// The `SELECT` clause for `SQL` subscription queries.
+///
+/// ```ebnf
+/// SELECT ( '*' | table '.' '*' )
+/// ```
+#[derive(Debug, Eq, PartialEq)]
+pub enum SqlProject {
+    /// An qualified `table.*`
+    QualifiedWildcard { table: Identifier },
+    /// An unqualified `SELECT *`
+    Wildcard,
+}
+
+/// The `ON` clause of `JOIN` for `SQL` subscription queries.
+///
+/// ```ebnf
+/// predicate_on
+///    = columnExpr '=' columnExpr
+///   | predicate_on AND predicate_on
+///  ;
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SqlOn {
+    And { lhs: Box<SqlOn>, rhs: Box<SqlOn> },
+    Eq { lhs: SqlColumn, rhs: SqlColumn },
+}
+
+/// The `[INNER] JOIN` clause for `SQL` subscription queries.
+///
+/// ```ebnf
+/// [INNER] JOIN table [ [AS] alias ] ON predicate_on
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SqlJoin {
+    Inner { rhs: SqlTable, on: SqlOn },
+}
+
+/// The `FROM` & `JOIN` clauses for `SQL` subscription queries.
+///
+/// ```ebnf
+/// FROM table [ [AS] alias ] [ [INNER] JOIN ]
+/// ```
+#[derive(Debug)]
+pub struct SqlFrom {
+    pub root: SqlTable,
+    pub joins: Vec<SqlJoin>,
+}
+
+impl SqlFrom {
+    pub fn new(root: SqlTable, joins: Vec<SqlJoin>) -> Self {
+        Self { root, joins }
+    }
+}
+
+impl From<(SqlTable, Vec<SqlJoin>)> for SqlFrom {
+    fn from(value: (SqlTable, Vec<SqlJoin>)) -> Self {
+        SqlFrom::new(value.0, value.1)
+    }
+}
+
+/// A valid expression for `SQL` subscription queries.
+///
+/// ```ebnf
+/// expr
+///     = literal
+///     | ident
+///     | column
+///     | expr op expr
+///     ;
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SqlExpr {
```

Comment:
This should have at least 4 variants: for literals, idents, field expressions, and binary expressions.

Reply:
The `sql parser` doesn't separate `ident` from `column` because is not possible to know without type checking, so neither I can add an `ident` here. I assume this is always a `SqlColumn`. If we say we can neither, then we need to change how `Identifier`  works, and remove `SqlColumn`

Reply:
Changed `SqlColumn` to `SqlIdentifier` so is more explicit.